File: goop/cmd/roll.go
// Author:  Niels A.D.
// Project: goop (https://github.com/nielsAD/goop)
// License: Mozilla Public License, v2.0

package cmd

import (
	"fmt"
	"math/rand"
	"strconv"

	"github.com/nielsAD/goop/gateway"
	"github.com/nielsAD/goop/goop"
)

// Roll a dice
type Roll struct{ Cmd }

// Execute command
func (c *Roll) Execute(t *gateway.Trigger, gw gateway.Gateway, g *goop.Goop) error {
	var max = int64(100)
	if len(t.Arg) > 0 {
		v, err := strconv.ParseInt(t.Arg[0], 0, 32)
		if err != nil {
			return t.Resp(err.Error())
		}
		max = v
	}
	if max <= 0 {
		return t.Resp("Pick a positive number larger than 0")
	}
	return t.Resp(fmt.Sprintf("%d", rand.Int63n(max)))
}

// Flip a coin
type Flip struct{ Cmd }

// Execute command
func (c *Flip) Execute(t *gateway.Trigger, gw gateway.Gateway, g *goop.Goop) error {
	var coin = "Heads"
	if rand.Float64() < 0.5 {
		coin = "Tails"
	}
	return t.Resp(coin)
}

File: plugins/roll.lua
-- Author:  Niels A.D.
-- Project: goop (https://github.com/nielsAD/goop)
-- License: Mozilla Public License, v2.0
--
-- Add roll/flip commands that can be used to roll a dice/flip a coin

defoptions({
    AccessTrigger = access.Voice, -- Access level required to trigger command
})

goop:AddCommand("roll", command(function(trig, _)
    if trig.User.Access < options.AccessTrigger then
        return nil
    end

    local max = 100
    if #trig.Arg > 0 then
        max = tonumber(trig.Arg[1])
        if max == nil then
            return trig.Resp("Cannot convert " .. trig.Arg[1] .. " to number")
        end
    end
    if max <= 0 then
        return trig.Resp("Pick a number above 0")
    end
    return trig.Resp(tostring(math.random(max)))

end))

goop:AddCommand("flip", command(function(trig, _)
    if trig.User.Access < options.AccessTrigger then
        return nil
    end

    local coin = "Heads"
    if math.random() < 0.5 then
        coin = "Tails"
    end
    return trig.Resp(coin)
end))
